### PR TITLE
support GPU tensor for `GridPatch` and `GridPatchDataset`

### LINKED ISSUE #6245: `GridPatchDataset` does not support GPU tensor as the data source
**Describe the bug**
Execute this script,
```
import numpy as np
import torch
from monai.data.dataloader import DataLoader
from monai.data.grid_dataset import GridPatchDataset, PatchIter

device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
print('Using device:', device)

images = [torch.arange(16, dtype=float, device=device).reshape(1, 4, 4), torch.arange(16, dtype=float, device=device).reshape(1, 4, 4)]
patch_iter = PatchIter(patch_size=(2, 2), start_pos=(0, 0))
ds = GridPatchDataset(data=images, patch_iter=patch_iter)

for item in DataLoader(ds, batch_size=2, shuffle=False, num_workers=0):
    np.testing.assert_equal(tuple(item[0].shape), (2, 1, 2, 2))
```
shows
QQ图片20230327220405

**To Reproduce**
see above

**Expected behavior**
no exception

**Environment**
```
================================
Printing MONAI config...
================================
MONAI version: 1.2.0rc1+22.g8eceabf2
Numpy version: 1.23.5
Pytorch version: 2.0.0+cu117
MONAI flags: HAS_EXT = False, USE_COMPILED = False, USE_META_DICT = False
MONAI rev id: 8eceabf281ab31ea4bda0ab8a6d2c8da06027e82
MONAI __file__: /mnt/d/codes/MONAI/monai/__init__.py

Optional dependencies:
Pytorch Ignite version: 0.4.11
ITK version: 5.3.0
Nibabel version: 5.0.1
scikit-image version: 0.20.0
Pillow version: 9.4.0
Tensorboard version: 2.12.0
gdown version: 4.6.4
TorchVision version: 0.15.1+cu117
tqdm version: 4.65.0
lmdb version: 1.4.0
psutil version: 5.9.4
pandas version: 1.5.3
einops version: 0.6.0
transformers version: 4.21.3
mlflow version: 2.2.2
pynrrd version: 1.0.0

For details about installing the optional dependencies, please visit:
    https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies


================================
Printing system config...
================================
System: Linux
Linux version: Ubuntu 22.04.2 LTS
Platform: Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Processor: x86_64
Machine: x86_64
Python version: 3.10.6
Process name: python
Command: ['python', '-c', 'import monai; monai.config.print_debug_info()']
Open files: []
Num physical CPUs: 6
Num logical CPUs: 12
Num usable CPUs: 12
CPU usage (%): [13.5, 2.7, 3.2, 3.7, 1.0, 0.9, 1.7, 1.6, 0.7, 0.5, 0.6, 0.7]
CPU freq. (MHz): 3600
Load avg. in last 1, 5, 15 mins (%): [8.4, 3.8, 2.2]
Disk usage (%): 68.9
Avg. sensor temp. (Celsius): UNKNOWN for given OS
Total physical memory (GB): 25.0
Available memory (GB): 24.2
Used memory (GB): 0.5

================================
Printing GPU config...
================================
Num GPUs: 1
Has CUDA: True
CUDA version: 11.7
cuDNN enabled: True
cuDNN version: 8500
Current device: 0
Library compiled for CUDA architectures: ['sm_37', 'sm_50', 'sm_60', 'sm_70', 'sm_75', 'sm_80', 'sm_86']
GPU 0 Name: NVIDIA GeForce GTX 1660
GPU 0 Is integrated: False
GPU 0 Is multi GPU board: False
GPU 0 Multi processor count: 22
GPU 0 Total memory (GB): 6.0
GPU 0 CUDA capability (maj.min): 7.5
```

**Additional context**
In an application scenario, we might need to get 2d slices from 3d (https://github.com/Project-MONAI/tutorials/blob/main/modules/2d_slices_from_3d_training.ipynb). When we move the data to GPU before `GridPatchDataset`  for acceleration (https://github.com/Project-MONAI/tutorials/blob/main/acceleration/fast_training_tutorial.ipynb), the function  `iter_patch` of `PatchIter` cannot handle GPU tensor correctly.  
  
My proposed PR will be linked later.

### PR DESCRIPTION
**Pull request by qingpeng9802**
Fixes #6245 .

### Description
support GPU tensor for
* `GridPatch`, `GridPatchd`, `RandGridPatch` and `RandGridPatchd`
* `GridPatchDataset`, `PatchIter`, `PatchIterd` and `iter_patch`

#### Changes
* Currently, `GridPatch`, `GridPatchDataset` and their siblings need to copy GPU tensor to numpy array (or not support GPU tensor) in their pipeline. This PR enables them to support end-to-end GPU operations.
*  Since the padding function `pad_nd` is used heavily, its robustness is improved. Some strange behavior introduced by https://github.com/Project-MONAI/MONAI/issues/6066 is fixed.
* In `GridPatch`, https://github.com/Project-MONAI/MONAI/issues/6082 might be fixed by this PR?
* `RandGridPatch`'s mode defaults to `None` (consistent with docs)

#### Need to Notice
* Default padding mode of `PatchIter`, `PatchIterd` and `iter_patch` are kept as `NumpyPadMode.WRAP`. However, `GridPatch` defaults to `None`. (might be inconsistent?)

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/transforms/croppad/functional.py [hunk 1]
```diff
@@ -148,23 +149,26 @@ def crop_or_pad_nd(img: torch.Tensor, translation_mat, spatial_size: tuple[int,
 
 
 def pad_func(
-    img: torch.Tensor, to_pad: tuple[tuple[int, int]], mode: str, transform_info: dict, kwargs
+    img: torch.Tensor, to_pad: tuple[tuple[int, int]], transform_info: dict, mode: str=PytorchPadMode.CONSTANT, **kwargs
```
Comment (wyli) [inline monai/transforms/croppad/functional.py:152]: it seems that this change is not needed by this PR, could you please confirm?

Comment (qingpeng9802) [reply monai/transforms/croppad/functional.py:152]: Since the last line here https://github.com/Project-MONAI/MONAI/blob/be3d13869d9e0060d17a794d97d528d4e4dcc1fc/monai/transforms/croppad/functional.py#L150-L164 mentioned the default padding mode is `PytorchPadMode.CONSTANT` (consistent with `pad_nd`'s signature), which make `mode` become a keyword parameter, the order of parameters needs to be changed. Would we keep this change or use the old `pad_func`'s signature?

Comment (wyli) [reply monai/transforms/croppad/functional.py:152]: I see, thanks, I think we can keep this change, also `mode` and `kwargs` are logically related, makes sense.

**Comment by qingpeng9802** [comment]
By checking the tests in CI, it seems `torch.argwhere` has some compatibility issues. From https://github.com/pytorch/pytorch/pull/64257, would it be better to replace `torch.argwhere` with `torch.nonzero` or a version-specified solution?

**Comment by wyli** [comment]
> By checking the tests in CI, it seems `torch.argwhere` has some compatibility issues. From [pytorch/pytorch#64257](https://github.com/pytorch/pytorch/pull/64257), would it be better to replace `torch.argwhere` with `torch.nonzero` or a version-specified solution?

thanks, let me skip the tests for this new feature for torch<1.11...

**Comment by wyli** [comment]
there are still some typing issues, I think you can debug with `./runtests.sh --mypy` locally. also please follow this page to sign the commit https://github.com/Project-MONAI/MONAI/pull/6246/checks

**Review by wyli — APPROVED** [review]
thanks, it looks good to me. @ericspod @Nic-Ma @KumoLiu if you have further comments I can create followup PRs.
